gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GcsJavaClientWriteChannel.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2023 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.encodeMetadata;
+import static com.google.storage.v2.ServiceConstants.Values.MAX_WRITE_CHUNK_BYTES;
+
+import com.google.cloud.WriteChannel;
+import com.google.cloud.hadoop.util.AbstractGoogleAsyncWriteChannel;
+import com.google.cloud.hadoop.util.AsyncWriteChannelOptions;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobWriteOption;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.io.ByteStreams;
+import com.google.protobuf.ByteString;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+
+/** Implements WritableByteChannel to provide write access to GCS via java-storage client */
+class GcsJavaClientWriteChannel extends AbstractGoogleAsyncWriteChannel<Boolean> {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final GoogleCloudStorageOptions storageOptions;
+  private final StorageResourceId resourceId;
+  private final AsyncWriteChannelOptions channelOptions;
+  private final Storage storage;
+  private WriteChannel writeChannel;
+  private final CreateObjectOptions createOptions;
+  private Boolean uploadSuccess = false;
+  // TODO: not supported as of now
+  // private final String requesterPaysProject;
+
+  public GcsJavaClientWriteChannel(
+      Storage storage,
+      GoogleCloudStorageOptions storageOptions,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      ExecutorService uploadThreadPool) {
+    super(uploadThreadPool, storageOptions.getWriteChannelOptions());
+    this.storage = storage;
+    this.storageOptions = storageOptions;
+    this.resourceId = resourceId;
+    this.createOptions = createOptions;
+    this.channelOptions = storageOptions.getWriteChannelOptions();
+  }
+
+  @Override
+  public void initialize() throws IOException {
+    final BlobId blobId =
```

Comment:
We generally do not use final for local variables. Here we are using final for some and not for some others.

Reply:
Will clean it.